File: python/tests/test_fbsi.py
"""Tests for the FBSI reference implementation.  Run with:  pytest python/tests"""

import numpy as np
import pytest

import fbsi
from fbsi import Cell, exterior_cell, face_weights, fbsi_fractions


def cap_fraction(d, radius):
    """Exact volume fraction of a sphere on the inner side (d < 0) of a plane."""
    h = np.clip(radius - np.asarray(d, dtype=float), 0.0, 2.0 * radius)
    return h * h * (3.0 * radius - h) / (4.0 * radius**3)


# ----------------------------------------------------------------- geometry
def test_box_and_tetrahedron_geometry():
    box = Cell.box([0, 0, 0], [2, 1, 1])
    assert box.volume == pytest.approx(2.0)
    assert np.allclose(box.centroid, [1.0, 0.5, 0.5])
    tet = Cell.tetrahedron([0, 0, 0], [1, 0, 0], [0, 1, 0], [0, 0, 1])
    assert tet.volume == pytest.approx(1.0 / 6.0)
    assert np.allclose(tet.centroid, [0.25, 0.25, 0.25])


def test_normals_point_outward_regardless_of_vertex_order():
    vertices = np.array([[0, 0, 0], [1, 0, 0], [0, 1, 0], [0, 0, 1]], dtype=float)
    faces = [[0, 1, 2], [0, 1, 3], [0, 2, 3], [1, 2, 3]]
    a = Cell.from_polyhedron(vertices, faces)
    b = Cell.from_polyhedron(vertices, [face[::-1] for face in faces])
    assert np.allclose(a.normals, b.normals)
    assert np.all(a.signed_distance(a.centroid) < 0.0)


def test_meshes_fill_the_box():
    for mesh, n_boundary in [(fbsi.hex_mesh(3), 6 * 9), (fbsi.tet_mesh(3), 6 * 9 * 2)]:
        assert mesh.volumes.sum() == pytest.approx(1.0)
        assert len(mesh.boundary_faces) == n_boundary


# ------------------------------------------------------------- face weight
def test_face_weight_limits_and_midpoint():
    w = face_weights(np.array([-2.0, -1.0, 0.0, 1.0, 2.0]), 1.0)
    assert np.allclose(w, [1.0, 1.0, 0.5, 0.0, 0.0])


def test_smoothstep_equals_spherical_cap_fraction():
    d = np.linspace(-1.2, 1.2, 101)
    assert np.allclose(face_weights(d, 1.0), cap_fraction(d, 1.0))


def test_face_weight_is_monotone_and_c1():
    d = np.linspace(-1.0, 1.0, 2001)
    w = face_weights(d, 1.0)
    assert np.all(np.diff(w) <= 0.0)
    slope = np.gradient(w, d)
    assert abs(slope[0]) < 1e-2 and abs(slope[-1]) < 1e-2  # zero slope at +-R


# --------------------------------------------------------------- fractions
def test_particle_inside_one_cell_goes_to_that_cell():
    cells = [Cell.box([0, 0, 0], [1, 1, 1]), Cell.box([1, 0, 0], [2, 1, 1])]
    assert np.allclose(fbsi_fractions([0.5, 0.5, 0.5], 0.3, cells), [1.0, 0.0])


@pytest.mark.parametrize("x", [0.75, 0.9, 1.0, 1.05, 1.2])
def test_single_face_crossing_is_exact(x):
    cells = [Cell.box([0, 0, 0], [1, 1, 1]), Cell.box([1, 0, 0], [2, 1, 1])]
    phi = fbsi_fractions([x, 0.5, 0.5], 0.3, cells)
    assert phi[0] == pytest.approx(cap_fraction(x - 1.0, 0.3), abs=1e-12)


def test_vertex_shared_by_eight_cells():
    cells = [Cell.box([i, j, k], [i + 1, j + 1, k + 1]) for i in (-1, 0) for j in (-1, 0) for k in (-1, 0)]
    assert np.allclose(fbsi_fractions([0, 0, 0], 0.3, cells), 1.0 / 8.0)


def test_partition_of_unity_on_tet_mesh():
    mesh = fbsi.tet_mesh(4)
    radius = 0.8 * mesh.mean_equivalent_radius()
    rng = np.random.default_rng(0)
    for center in rng.uniform(radius, 1 - radius, size=(50, 3)):
        phi = fbsi_fractions(center, radius, [mesh.cells[i] for i in mesh.candidates(center, radius)])
        assert np.all(phi >= 0.0)
        assert phi.sum() == pytest.approx(1.0)


def test_fractions_change_continuously():
    mesh = fbsi.tet_mesh(3)
    radius = 0.8 * mesh.mean_equivalent_radius()
    index = np.arange(len(mesh))
    previous = None
    for x in np.linspace(0.3, 0.7, 400):
        center = np.array([x, 0.41, 0.52])
        phi = np.zeros(len(mesh))
        candidates = mesh.candidates(center, radius)
        phi[index[candidates]] = fbsi_fractions(center, radius, [mesh.cells[i] for i in candidates])
        if previous is not None:
            assert np.max(np.abs(phi - previous)) < 0.05  # step is 1e-3, no jumps
        previous = phi


def test_error_when_no_candidate_overlaps():
    with pytest.raises(ValueError):
        fbsi_fractions([5.0, 5.0, 5.0], 0.1, [Cell.box([0, 0, 0], [1, 1, 1])])


# -------------------------------------------------------------- boundaries
@pytest.mark.parametrize("x", [0.06, 0.03, 0.0, -0.03])
def test_exterior_cell_captures_volume_outside_a_wall(x):
    mesh = fbsi.hex_mesh(5)
    radius, center = 0.08, np.array([x, 0.31, 0.49])
    interior = [mesh.cells[i] for i in mesh.candidates(center, radius)]
    phi = fbsi_fractions(center, radius, interior + mesh.exterior_cells(center, radius))
    assert phi[: len(interior)].sum() == pytest.approx(cap_fraction(-x, radius), abs=1e-9)


def test_exterior_cell_faces():
    ghost = exterior_cell([[0, 0, 0], [0, 1, 0], [0, 1, 1], [0, 0, 1]], interior_point=[0.5, 0.5, 0.5])
    assert ghost.exterior and ghost.n_faces == 5
    assert ghost.contains([-3.0, 0.5, 0.5])      # behind the wall
    assert not ghost.contains([0.5, 0.5, 0.5])   # inside the domain


# -------------------------------------------------------------- accuracy
def test_accuracy_against_direct_numerical_reference():
    mesh = fbsi.tet_mesh(5)
    radius = 0.8 * mesh.mean_equivalent_radius()
    rng = np.random.default_rng(3)
    err_fbsi, err_iso = [], []
    for center in rng.uniform(radius, 1 - radius, size=(40, 3)):
        cells = [mesh.cells[i] for i in mesh.candidates(center, radius)]
        reference = fbsi.direct_numerical_fractions(center, radius, cells, n_per_dim=48)
        err_fbsi.append(np.abs(fbsi_fractions(center, radius, cells) - reference).sum())
        err_iso.append(np.abs(fbsi.isotropic_fractions(center, radius, cells) - reference).sum())
    assert np.mean(err_fbsi) < 0.5 * np.mean(err_iso)
    assert np.mean(err_fbsi) < 0.15


def test_solid_volume_fraction_conserves_particle_volume():
    mesh = fbsi.hex_mesh(4)
    radius = 0.05
    rng = np.random.default_rng(1)
    centers = rng.uniform(radius, 1 - radius, size=(30, 3))
    v_p = 4.0 / 3.0 * np.pi * radius**3
    per_particle = []
    for center in centers:
        index = mesh.candidates(center, radius)
        per_particle.append((index, fbsi_fractions(center, radius, [mesh.cells[i] for i in index])))
    phi_cells = fbsi.solid_volume_fraction(per_particle, [v_p] * len(centers), mesh.volumes)
    assert np.sum(phi_cells * mesh.volumes) == pytest.approx(len(centers) * v_p)

File: cpp/tests/test_fbsi.cpp
// Tests for fbsi.hpp (no test framework needed). Run with: ctest --test-dir build

#include <fbsi/fbsi.hpp>

#include "simple_mesh.hpp"

#include <cmath>
#include <cstdio>
#include <random>
#include <vector>

using fbsi::Cell;
using fbsi::Vec3;

static int failures = 0;

#define CHECK(condition)                                                        \
    do {                                                                        \
        if (!(condition)) {                                                     \
            std::printf("FAILED: %s  (%s:%d)\n", #condition, __FILE__, __LINE__); \
            ++failures;                                                         \
        }                                                                       \
    } while (0)

static bool near(double a, double b, double tol = 1e-12) { return std::abs(a - b) <= tol; }

// Exact volume fraction of a sphere on the inner side (d < 0) of a plane.
static double cap_fraction(double d, double radius) {
    const double h = std::fmin(std::fmax(radius - d, 0.0), 2.0 * radius);
    return h * h * (3.0 * radius - h) / (4.0 * radius * radius * radius);
}

static double sum(const std::vector<double>& v) {
    double s = 0.0;
    for (double x : v) s += x;
    return s;
}

static void test_geometry() {
    const Cell box = fbsi::make_box({0, 0, 0}, {2, 1, 1});
    CHECK(near(box.volume, 2.0));
    CHECK(near(box.centroid.x, 1.0) && near(box.centroid.y, 0.5) && near(box.centroid.z, 0.5));
    const Cell tet = fbsi::make_tetrahedron({0, 0, 0}, {1, 0, 0}, {0, 1, 0}, {0, 0, 1});
    CHECK(near(tet.volume, 1.0 / 6.0));
    CHECK(fbsi::contains(tet, tet.centroid));
    CHECK(!fbsi::contains(tet, {1, 1, 1}));
}

static void test_face_weight() {
    CHECK(near(fbsi::face_weight(-2.0, 1.0), 1.0));
    CHECK(near(fbsi::face_weight(0.0, 1.0), 0.5));
    CHECK(near(fbsi::face_weight(2.0, 1.0), 0.0));
    for (double d = -1.2; d <= 1.2; d += 0.01) CHECK(near(fbsi::face_weight(d, 1.0), cap_fraction(d, 1.0)));
}

static void test_two_cells() {
    const Cell a = fbsi::make_box({0, 0, 0}, {1, 1, 1});
    const Cell b = fbsi::make_box({1, 0, 0}, {2, 1, 1});
    std::vector<double> phi;
    fbsi::fbsi_fractions({0.5, 0.5, 0.5}, 0.3, {&a, &b}, phi);
    CHECK(near(phi[0], 1.0) && near(phi[1], 0.0));
    for (double x : {0.75, 0.9, 1.0, 1.05, 1.2}) {
        fbsi::fbsi_fractions({x, 0.5, 0.5}, 0.3, {&a, &b}, phi);
        CHECK(near(phi[0], cap_fraction(x - 1.0, 0.3)));  // single face: exact
    }
    const double total = fbsi::fbsi_fractions({5, 5, 5}, 0.1, {&a, &b}, phi);
    CHECK(total == 0.0);  // no overlap is reported, not hidden
}

static void test_eight_cells_at_vertex() {
    std::vector<Cell> cubes;
    for (int i : {-1, 0})
        for (int j : {-1, 0})
            for (int k : {-1, 0}) cubes.push_back(fbsi::make_box({1.0 * i, 1.0 * j, 1.0 * k}, {i + 1.0, j + 1.0, k + 1.0}));
    std::vector<const Cell*> cells;
    for (const Cell& c : cubes) cells.push_back(&c);
    std::vector<double> phi;
    fbsi::fbsi_fractions({0, 0, 0}, 0.3, cells, phi);
    for (double v : phi) CHECK(near(v, 0.125));
}

static void test_partition_of_unity_and_accuracy() {
    const simple_mesh::Mesh mesh = simple_mesh::tet_mesh(5);
    const double radius = 0.8 * mesh.mean_equivalent_radius();
    std::mt19937_64 rng(3);
    std::uniform_real_distribution<double> uniform(radius, 1.0 - radius);
    double err = 0.0;
    const int n = 50;
    std::vector<double> phi;
    for (int p = 0; p < n; ++p) {
        const Vec3 c{uniform(rng), uniform(rng), uniform(rng)};
        const auto cells = mesh.candidates(c, radius);
        fbsi::fbsi_fractions(c, radius, cells, phi);
        CHECK(near(sum(phi), 1.0, 1e-12));
        const auto ref = fbsi::direct_numerical_fractions(c, radius, cells, 64);
        for (std::size_t i = 0; i < phi.size(); ++i) {
            CHECK(phi[i] >= 0.0);
            err += std::abs(phi[i] - ref[i]);
        }
    }
    CHECK(err / n < 0.15);  // mean L1 error per particle
}

static void test_wall() {
    const simple_mesh::Mesh mesh = simple_mesh::hex_mesh(5);
    const double radius = 0.08;
    std::vector<double> phi;
    for (double x : {0.06, 0.03, 0.0, -0.03}) {
        const Vec3 c{x, 0.31, 0.49};
        auto cells = mesh.candidates(c, radius);
        const std::size_t n_interior = cells.size();
        const std::vector<Cell> ghosts = mesh.exterior_cells(c, radius);
        for (const Cell& g : ghosts) cells.push_back(&g);
        fbsi::fbsi_fractions(c, radius, cells, phi);
        double inside = 0.0;
        for (std::size_t i = 0; i < n_interior; ++i) inside += phi[i];
        CHECK(near(inside, cap_fraction(-x, radius), 1e-9));
    }
}

int main() {
    test_geometry();
    test_face_weight();
    test_two_cells();
    test_eight_cells_at_vertex();
    test_partition_of_unity_and_accuracy();
    test_wall();
    if (failures == 0) std::printf("All FBSI tests passed.\n");
    return failures == 0 ? 0 : 1;
}
